### PR TITLE
Performance: reduce calls to kubernetes api-server

### PR DESCRIPTION
**Pull request by emouawad**
## Proposed change

⚠️ Note: RBAC: A ClusterRole / ClusterRoleBinding is required now when gateway: true - whereas a namespaced RBAC RoleBinding was enough.

Closes #5830 
Closes #5969

This is a performance enhancement that switches from looping on Kubernetes namespaces then calling get CRD for each namespace to using 1 API call to list HttpRoute CRD on all the cluster.

For a cluster that has 100+ namespaces this is noticeable as the CPU / Memory in the older implementation was 3-4 times greater and homepage was lagging / taking longer to respond.

image

## Type of change

Enhancement (non-breaking change which fixes performance on cluster with lots of namespaces)

## Checklist:

- No configuration change is needed
- RBAC: A ClusterRole / ClusterRoleBinding is required now when `gateway: true` - whereas a namespaced RBAC RoleBinding was enough.
- Unit tests updated and working
- Tested on a live cluster that had the previous issue

### CODE REVIEW

**Comment by emouawad** [comment]
@shamoon Can you look at this?
Automation closed it but i believe this is an important performance enhancement.
I fixed the linting issue in my branch, if this PR is reponed i can continue working on it to pass the requirements.
Thanks

**Comment by shamoon** [comment]
You triggered all kinds of AI slop flags, including committing with different users, bypassing the issue template, etc.

File: src/utils/kubernetes/httproute-list.js [hunk 1]
```diff
@@ -7,49 +7,29 @@ const logger = createLogger("httproute-list");
...
+    const crd = kc.makeApiClient(CustomObjectsApi);
+
+    const httpRoutes = await crd
+      .listClusterCustomObject({
+        group: HTTPROUTE_API_GROUP,
+        version: HTTPROUTE_API_VERSION,
+        plural: "httproutes",
+      })
+      .then((response) => {
+        logger.debug("Retrieved httproutes: %d", response?.items?.length ?? 0);
```
Comment (shamoon) [inline src/utils/kubernetes/httproute-list.js]:
```suggestion
```

noise

File: src/utils/kubernetes/httproute-list.js [hunk 2]
```diff
@@ -7,49 +7,29 @@ const logger = createLogger("httproute-list");
...
+        return response?.items;
```
Comment (shamoon) [inline src/utils/kubernetes/httproute-list.js]:
```suggestion
        return response?.items ?? [];
```

this will cause a crash later

**Comment by emouawad** [comment]
> You triggered all kinds of AI slop flags, including committing with different users, bypassing the issue template, etc.

Ah sorry :( my first contribution ever to any open source project; lol thanks for cutting me some slack!

Can you help me out - what is needed for the anti-slop check?

File: src/utils/kubernetes/httproute-list.test.js [hunk 3]
```diff
@@ -87,20 +74,4 @@ describe("utils/kubernetes/httproute-list", () => {
     expect(logger.error).toHaveBeenCalled();
     expect(logger.debug).toHaveBeenCalled();
   });
-
-  it("skips namespaces whose httproute queries fail", async () => {
```
Comment (shamoon) [inline src/utils/kubernetes/httproute-list.test.js:91]: Deleting a test with no explanation is an obvious red flag. Do you know what it did? What removing it does, or more specifically what this AI-generated PR changes about it?

**Review by shamoon — CHANGES_REQUESTED** [review]

File: src/utils/kubernetes/httproute-list.test.js (hunk 3 above)
Comment (emouawad) [reply src/utils/kubernetes/httproute-list.test.js:91]: Yes  - this test is no longer needed.

Previous implementation was looping on each namespace: n namespaces meant n api calls to kubernetes api server 

This implementation isn't using namespaces at all therefore this case is irrelevant as a test

Comment (shamoon) [reply src/utils/kubernetes/httproute-list.test.js:91]: Agreed the test can't survive the refactor, my point is about what it was guarding. This changes behavior for setups granting httproute access via namespaced RoleBindings instead of a ClusterRoleBinding: they go from partial results to an empty list. Yes the documented setup uses a ClusterRole, so that might be an acceptable trade off for the performance improvement but 'no configuration change is needed' isnt really accurate, and should at least be noted in the PR

**Comment by shamoon** [comment]
> Can you help me out - what is needed for the anti-slop check?

Committing as yourself, not as some weird other presumably LLM user, using the project's issue template, all of the basic etiquette that worked with OSS for many years before LLMs started generating... you know what.

File: src/utils/kubernetes/httproute-list.test.js (hunk 3 above)
Comment (emouawad) [reply src/utils/kubernetes/httproute-list.test.js:91]: Fair point, note this change aligns with the traefik implementation which is using listClusterCustomObject and as you mentioned the docs mention a ClusterRole/Binding

I will update my PR to reflect the RBAC change and update the docs?(https://github.com/gethomepage/homepage/blob/dev/docs/installation/k8s.md).

Comment (shamoon) [reply src/utils/kubernetes/httproute-list.test.js:91]: Since it’s probably an edge case just noting it in the PR will suffice

Comment (emouawad) [reply src/utils/kubernetes/httproute-list.test.js:91]: I have update the PR description as suggested

> RBAC: A ClusterRole / ClusterRoleBinding is required now when `gateway: true` - whereas a namespaced RBAC RoleBinding was enough.

Let me know if anything else is still needed - thanks

**Review by shamoon — APPROVED** [review]
